Add "Focus Pointer" Method to the 3D Editor Plugin

priahoud opened the pull request:
Useful to quickly navigate complex 3D scenes by focusing the viewport to the clicked position on a surface.
Can be activated by command with a keyboard shortcut, configurable in EditorSettings/Shortcuts/Focus Pointer.

Closes godotengine/godot-proposals#910

[capture.webm](https://github.com/user-attachments/assets/c7f00bc8-5757-4403-bdad-cbe1fd8b2e23)

<img width="808" height="629" alt="shortcuts" src="https://github.com/user-attachments/assets/6da224be-fbe2-44ad-9fb5-e541a6c4b8eb" />

Calinou reviewed (commented):
Tested locally, it works as expected.

However, if you orbit while holding <kbd>Alt</kbd>, it will only focus on the point when you release the mouse button. The orbiting operation will be relative to the old position, then it'll snap to the new position after you release the mouse.

I think it should set the new point of origin for orbiting immediately when you click. It would likely match user expectations better as I've understood from the proposal.

priahoud commented:
Thank you for the Review!

> However, if you orbit while holding Alt, it will only focus on the point when you release the mouse button. The orbiting operation will be relative to the old position, then it'll snap to the new position after you release the mouse.
>
> I think it should set the new point of origin for orbiting immediately when you click. It would likely match user expectations better as I've understood from the proposal.

That was intentional because that is also the default behaviour in Blender and Unity. This is because both differentiate between mouse clicking and dragging, which makes it necessary to fire the click on mouse up after checking for a drag on mouse down.

In Blender, holding `MMB down` is orbit, holding `Alt`+`MMB down` is switch orthographic views and clicking `Alt`+`MMB up` is focus pointer.

In Unity, holding `MMB down` is panning, holding `Alt`+`MMB down` is orbit and clicking `MMB up` is focus pointer.

I use both of them regularly and this is the behaviour I would have expected. Of course that doesn't mean we should blindly follow what they have done. I did a quick test with focus on `Alt`+`MMB down` and it did feel a bit weird at first, but was quite pleasant to use after a while.

Though I now also realize that hardcoding `Alt`+`MMB` is probably not a good idea because Godot supports [Navigation Schemes](https://docs.godotengine.org/en/stable/tutorials/3d/introduction_to_3d.html#navigating-the-3d-environment) and the `Tablet/Trackpad` scheme does use `Alt`+`MMB` for orbit. So it might still be necessary to make that distinction between mouse down and up. Will look into that.

Calinou reviewed (commented):
Tested again, I think it works pretty well now. I've tried all navigation schemes and was wondering why **Tablet/Trackpad** used <kbd>Ctrl + Alt + Middle mouse button</kbd> instead of just <kbd>Alt + Middle mouse button</kbd> though.

Tablet/Trackpad is a special case when it comes to navigation, as it will work without holding down any mouse buttons. Just hold down <kbd>Alt</kbd> and move the mouse to orbit (<kbd>Ctrl</kbd> for zoom, <kbd>Shift</kbd> for pan).

Calinou commented:
Something odd is happening with the latest revision of this PR: I can use the feature in the first scene I'm opening (or restoring after reopening the editor), but if I switch scene tabs, I can't use it in the other tab. Even if I switch back to the first tab I was using, I still can't make use of the feature; nothing happens when I use <kbd>Alt + Middle mouse button</kbd> until I restart the editor.

priahoud commented:
> Something odd is happening with the latest revision of this PR: I can use the feature in the first scene I'm opening (or restoring after reopening the editor), but if I switch scene tabs, I can't use it in the other tab. Even if I switch back to the first tab I was using, I still can't make use of the feature; nothing happens when I use Alt + Middle mouse button until I restart the editor.
> 
> It's possible that earlier revisions of this PR were also affected, but I didn't notice that until now.

That is really odd. I tried to reproduce this with two scenes opened in different tabs on my local machine, but everything works as expected here. Will investigate further.

priahoud commented:
Ok, this is really weird!

First, apologies for the delay. After not being able to reproduce the error on my machine at home (Fedora on Wayland) and not having access to a Windows machine, I spent some time trying to prepare a Windows VM that I could use to test MinGW builds on, but it seems like current VM graphic drivers are not enough to successfully run Godot in a Windows VM.

So I decided to try some things on my Windows machine at work.
- MinGW build cross compiled on my Fedora machine at home - **cannot reproduce**
- Visual Studio build compiled on my Windows machine at work - **cannot reproduce**
- LLVM-MinGW build compiled on my Windows machine at work - **cannot reproduce**
- [Windows Build Artifact](https://github.com/godotengine/godot/actions/runs/16450303157/artifacts/3589638391) downloaded from this PR on my Windows machine at work - **does  reproduce**

Pretty puzzled about this one too, because besides setting a new cursor position, no state is changed here. The only thing I could see messing with other states are the breaks in the switch statements, which could possibly prevent important code further down from managing state. A build without breaks also runs fine here on my home machine, so I will try to push that change and see if that does resolve the problems on Windows.

My main problem is still that I am unable to compile any build that reproduces this error. As a last resort I will try to use the [official build-containers](https://github.com/godotengine/build-containers) and see if that makes any difference.

If all else fails, I will just rip out the navigation scheme code and either only leave the keyboard shortcut, or add the hardcoded `Alt`+`MMB` again.

Calinou commented:
> but it seems like current VM graphic drivers are not enough to successfully run Godot in a Windows VM.

Try disabling 3D acceleration in the VM and installing [mesa-dist-win](https://github.com/pal1000/mesa-dist-win) system-wide in the VM. This should allow the Compatibility renderer to work, albeit slowly.

priahoud committed:
Add "Focus Pointer" Method to the 3D Editor Plugin
Useful to quickly navigate complex 3D scenes by focusing the viewport to the clicked position on a surface.
Can be activated by command with a keyboard shortcut, configurable in EditorSettings/Shortcuts/Focus Pointer.

Closes godotengine/godot-proposals#910

KoBeWi commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -5778,6 +5798,7 @@ Node3DEditorViewport::Node3DEditorViewport(Node3DEditor *p_spatial_editor, int p
 
 	view_display_menu->get_popup()->add_separator();
 	view_display_menu->get_popup()->add_shortcut(ED_GET_SHORTCUT("spatial_editor/focus_origin"), VIEW_CENTER_TO_ORIGIN);
+	view_display_menu->get_popup()->add_shortcut(ED_GET_SHORTCUT("spatial_editor/focus_pointer"), VIEW_CENTER_TO_POINTER);
```
This option does not make much sense. You can't select it and point at an object at the same time. `_sinput()` already handles the shortcut.

priahoud replied:
Hey, thanks for the review!

> This option does not make much sense.

Except for discoverability, but I see your point.

> You can't select it and point at an object at the same time.

That is true.

> `_sinput()` already handles the shortcut.

After commenting that line out, the keyboard shortcut for focus pointer only works once until the ui gets "ticked" again in any way (the little wheel on the top right of the editor moves a bit) . Left or right clicking  in an empty area of the scene or some other widget in the editor, hovering the mouse over another area than the active one, or by just switching to another open application and back to Godot.

Because of the similarity to #108983, I did a quick cherry pick onto the commit before 6253814 to see if it is related, but that doesn't seem to be the case.

Any ideas?

KoBeWi replied:
> Except for discoverability, but I see your point.

Discoverability is important, but it should not be added as non-functional option. We could e.g. add a toggleable help panel that lists navigation shortcuts. It can be done in a separate PR.

> After commenting that line out, the keyboard shortcut for focus pointer only works once until the ui gets "ticked" again in any way

??
What I meant is that you can replace `_menu_option(VIEW_CENTER_TO_POINTER);` with `focus_pointer();`. The menu option is not needed for the shortcut to function, though it seems like you only added it for discovearibility?

Not sure about the bug you mentioned, I can't reproduce it.

priahoud commented:
Ok, with #109029 merged we are back in business, but there is still an issue with the way navigation schemes are evaluated. They basically don't differentiate between all possible combinations of modifier keys, leading to situations where focusing to the pointer also interferes with another navigation function.

For example in the Godot scheme, orbit navigation is configured to be `MMB` without modifier keys. `Alt`+`MMB` would be a perfect shortcut for focus pointer (as that is also the shortcut in blender), but the system doesn't take non-assigned modifier keys into account. As far as i can see, it just prioritizes the shortcut with the most defined modifiers. As a result, both functions would be activated at the same time when pressing `Alt`+`MMB`.  The same is true for the Maya scheme and panning.

Generally, a choice of a shortcut combo for focus pointer in the ui does not make much sense when half of them interfere with other ones. I am currently checking to see if that limitation can be worked around but until then, I have removed the navigation scheme parts for now.

KoBeWi commented:
Mouse buttons are not handled using shortcuts, this is probably hard-coded in the editor somewhere. You can modify the code to properly take modifiers into account.
